bluetooth: host: add frame space update support

sean-madigan opened the pull request:
This commit adds support for the frame space update feature to the bluetooth host.

This is mainly just a wrapper around the frame space update HCI command and event.

Fix BT_LE_LOCAL_SUPPORTED_FEATURES_SIZE define, as this was not working correctly without brackets around
CONFIG_BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE.

hermabe commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -77,10 +77,11 @@ extern "C" {
-	(BT_HCI_LE_BYTES_PAGE_0_FEATURE_PAGE +                                                     \
-	 COND_CODE_1(CONFIG_BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE,                                \
-		CONFIG_BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE * BT_HCI_LE_BYTES_PER_FEATURE_PAGE,  \
+#define BT_LE_LOCAL_SUPPORTED_FEATURES_SIZE                         \
```
If this is a bugfix it should probably be in a separate commit.

hermabe commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -273,6 +273,71 @@ struct bt_conn_le_read_all_remote_feat_complete {
+	 *  BT_CONN_LE_FRAME_SPACE_TYPES_MASK_* for values.
+	 */
+	uint16_t spacing_types;
+	/** Minimum frame space in microseconds. */
+	uint16_t frame_space_min;
+	/** Maximum frame space in microseconds. */
+	uint16_t frame_space_max;
```
Maybe add a reference to the HCI spec for the valid range for these values?

hermabe commented on subsys/bluetooth/Kconfig:
```diff
@@ -117,6 +117,7 @@ config BT_CONN_TX
 
 config BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE
 	int "Maximum supported feature page"
+	default 1 if BT_FRAME_SPACE_UPDATE
```
Would something break if the application sets this config to 0 while `BT_FRAME_SPACE_UPDATE` is enabled? If so, should this config be promptless or update the `range` as well if `BT_FRAME_SPACE_UPDATE`?

sean-madigan replied:
That's a good point, thanks

hermabe commented on subsys/bluetooth/host/conn.c:
```diff
@@ -3355,6 +3355,75 @@ int bt_conn_le_read_all_remote_features(struct bt_conn *conn, uint8_t pages_requ
+{
+	struct bt_conn_cb *callback;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn_cbs, callback, _node) {
```
There is a conflict with https://github.com/zephyrproject-rtos/zephyr/pull/93703 here. Whichever is merged last will need to rebase and update this.

hermabe replied:
This has been merged now

hermabe commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1849,6 +1849,33 @@ static void le_read_all_remote_feat_complete(struct net_buf *buf)
+
+	params.status = evt->status;
+
+	if (params.status == BT_HCI_ERR_SUCCESS) {
```
The docs in the header says:
> The remaining parameters will be unchanged if status is not @ref BT_HCI_ERR_SUCCESS.

But the fields are not set so they will contain garbage. Should the api description be updated to say the rest of the fields are invalid in case of an HCI error?

Alternatively we would need to store the current frame space values somewhere in host and populate `params` with those, but that seems inefficient.

hermabe commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -3352,6 +3387,43 @@ static int cmd_read_all_remote_features(const struct shell *sh, size_t argc, cha
+	const struct bt_conn_le_frame_space_update_param params = {
+		.frame_space_min = shell_strtoul(argv[1], 10, &err),
+		.frame_space_max = shell_strtoul(argv[2], 10, &err),
+		.phys = shell_strtoul(argv[3], 10, &err),
+		.spacing_types = shell_strtoul(argv[4], 10, &err),
```
Does this require the user to enter the bitmasks as base-10 numbers? It would be more intuitive for the user to enter them as hex, I would think.

sean-madigan committed:
bluetooth: host: add frame space update support

This commit adds support for the frame space update feature
to the bluetooth host.

This is mainly just a wrapper around the frame space update
HCI command and event.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

hermabe commented on subsys/bluetooth/Kconfig:
```diff
@@ -115,14 +115,20 @@ config BT_CONN_TX
 
+config BT_LE_LOCAL_MINIMUM_REQUIRED_FEATURE_PAGE
+	int
+	default 1 if BT_FRAME_SPACE_UPDATE
```
Should there be a `default 0` as a fallback here incase the application for whatever reason enables `BT_LE_EXTENDED_FEAT_SET` but none of the features that requires it?

```suggestion
	default 1 if BT_FRAME_SPACE_UPDATE
	default 0
```

sean-madigan replied:
Great point

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2815,6 +2818,26 @@ struct bt_hci_cp_le_cs_remove_config {
 
 #define BT_HCI_OP_LE_CS_TEST_END BT_OP(BT_OGF_LE, 0x0096) /* 0x2096 */
 
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_PHY_1M_MASK    BIT(0)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_PHY_2M_MASK    BIT(1)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_PHY_CODED_MASK BIT(2)
+
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_ACL_CP_MASK BIT(0)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_ACL_PC_MASK BIT(1)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_SPACING_TYPE_MCES_MASK       BIT(2)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_CIS_MASK    BIT(3)
+#define BT_HCI_OP_LE_FRAME_SPACE_UPD_SPACING_TYPE_MSS_CIS_MASK    BIT(4)
```
They're also quite long, leading to hard to read code. Would it hurt readability a lot if you e.g. shortened frame_space_update in all places to fsu?

jhedberg commented on tests/bluetooth/shell/testcase.yaml:
```diff
@@ -40,6 +40,12 @@ tests:
+    extra_configs:
+      - CONFIG_BT_FRAME_SPACE_UPDATE=y
+      - CONFIG_BT_LE_EXTENDED_FEAT_SET=y
+      - CONFIG_BT_LL_SW_SPLIT=n
```
Hmm, though I can see the same pattern repeating in the other tests above this one. Can you explain the reasoning for this?

alwa-nordic replied:
I believe it's done to make `HAS_BT_CTLR=n`, which will allow the features to be build-tested, even though `bt_enable()` will obviously fail since there is no controller in the system.

alwa-nordic reviewed (changes requested):
Needs fix for the reference errors that Copilot points out. It's a good idea to do define symbols for the constants like it asks about as well.

alwa-nordic commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -2815,6 +2818,30 @@ struct bt_hci_cp_le_cs_remove_config {
 
 #define BT_HCI_OP_LE_CS_TEST_END BT_OP(BT_OGF_LE, 0x0096) /* 0x2096 */
 
+#define BT_HCI_LE_FRAME_SPACE_UPD_PHY_1M_MASK    BIT(0)
+#define BT_HCI_LE_FRAME_SPACE_UPD_PHY_2M_MASK    BIT(1)
+#define BT_HCI_LE_FRAME_SPACE_UPD_PHY_CODED_MASK BIT(2)
+
+#define BT_HCI_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_ACL_CP_MASK BIT(0)
+#define BT_HCI_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_ACL_PC_MASK BIT(1)
+#define BT_HCI_LE_FRAME_SPACE_UPD_SPACING_TYPE_MCES_MASK       BIT(2)
+#define BT_HCI_LE_FRAME_SPACE_UPD_SPACING_TYPE_IFS_CIS_MASK    BIT(3)
+#define BT_HCI_LE_FRAME_SPACE_UPD_SPACING_TYPE_MSS_CIS_MASK    BIT(4)
+
+#define BT_HCI_LE_FRAME_SPACE_UPD_INITIATOR_LOCAL_HOST       (0)
+#define BT_HCI_LE_FRAME_SPACE_UPD_INITIATOR_LOCAL_CONTROLLER (1)
+#define BT_HCI_LE_FRAME_SPACE_UPD_INITIATOR_PEER             (2)
```
I agree. We should be consistent so this should be fixed in this PR.

alwa-nordic replied:
Out of scope of this PR definitely, but I suggest that we start with an approved abbreviation and synonyms documentation file of some kind. Ideally machine readable. I would love an editor extension that can look up an abbreviation for me.

sean-madigan committed:
bluetooth: host: Fix BT_LE_LOCAL_SUPPORTED_FEATURES_SIZE define

Fix BT_LE_LOCAL_SUPPORTED_FEATURES_SIZE define, as this was
not working correctly without brackets around
CONFIG_BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE.

Also fix CONFIG_BT_LE_MAX_LOCAL_SUPPORTED_FEATURE_PAGE help,
as this was talking about remote featues, where this only
applies for local features.

Also update for new CONFIG_BT_CONN_DYNAMIC_CALLBACKS kconfig

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>
